### ISSUE TITLE
Memory unsafety after deepcopy of plan

### ISSUE THREAD

**Issue by kbarros** [JuliaMath/FFTW.jl#261]
The following will segfault on my machine:

```
using FFTW

function make_plan()
    plan = FFTW.plan_fft(zeros(10))
    return deepcopy(plan)
end

plan = make_plan()
GC.gc() # presumably C datastructures are freed in this step, even though they remain live
plan * ones(10) # segfault
```

Since FFTW plans are immutable, a quick fix might be to have `deepcopy` return the original plan,
```
Base.deepcopy(plan::FFTW.cFFTWPlan) = plan
# ... repeat for all FFTW plan types
```

Alternatively, attempting to `deepcopy` an FFTW plan could throw an error. In any case, it seems that memory unsafety should be avoided.

Tested on FFTW v1.5.0 with Julia 1.9 beta 4:
```
julia> versioninfo()
Julia Version 1.9.0-beta4
Commit b75ddb787ff (2023-02-07 21:53 UTC)
Platform Info:
  OS: macOS (arm64-apple-darwin21.5.0)
  CPU: 8 × Apple M1 Pro
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-14.0.6 (ORCJIT, apple-m1)
  Threads: 1 on 6 virtual cores
```

**Comment by stevengj** [comment]
Yeah, I would just return the original plan object.

See also https://github.com/FFTW/fftw3/issues/106 which would be required for a "true" `deepcopy`.

**Comment by kbarros** [comment]
Could the mutable `pinv` field create issues with thread safety if the "copied" plan is attempted to be used simultaneously on another thread?

**Comment by stevengj** [comment]
Yes, seems like there could be a race condition if two threads try to update `pinv` simultaneously?    Though I guess we could add a lock.

**Comment by kbarros** [comment]
I think I agree that the easy thing for now is to fix the crash by adding the methods:
```
Base.deepcopy(plan::FFTW.cFFTWPlan) = plan
# ...
```

Later you could consider adding a lock, or a "true" `deepcopy` into FFTW itself, to provide thread safety.

**Comment by kbarros** [comment]
After pondering this a bit more, perhaps implementing `deepcopy` as a shallow copy is not great idea. Mainly, there are the potential thread safety issues. But also, `deepcopy` is a surprisingly dangerous function, especially in conjunction with C libraries, and there is no warning about this in the Julia docs. An error along the lines "`deepcopy` not implemented for FFTW plans" is factually correct (for the moment) and could also be helpful for keeping people within 'safe Julia'.

**Comment by kbarros** [comment]
Now that https://github.com/FFTW/fftw3/issues/106 is closed, is it possible to add `deepcopy` in a thread-safe way? The motivation would be sharing of plans across threads.

**Comment by kbarros** [comment]
Gentle ping. Can `Base.deepcopy(plan::FFTW.cFFTWPlan)` and friends be implemented now that `fftw_clone` is available? At the moment, calling Julia's `deepcopy` function on an FFTW plan will create nondeterministic memory corruption (triggered upon garbage collection).

**Comment by stevengj** [comment]
We need to build a new FFTW release.

**Comment by KristofferC** [comment]
I think this is not too unexpected. You really shouldn't use deepcopy and in combination with C interop it's very much asking for trouble. What's the reason for calling deepcopy, what property from it do you want?

**Comment by kbarros** [comment]
> What's the reason for calling deepcopy, what property from it do you want?

Sorry, let me clarify the two reasons why I think this issue remains important:

1. The original motivation was to make my simulations thread-safe. As previously discussed, FFTW plans have some cached state, and therefore shouldn't be reused across threads. So to run simulations on multiple threads, it would be convenient to call `copy(plan)` or equivalent. Since that's not implemented for FFTW plans, a user might naively reach for `deepcopy`...
2. As you point out, `deepcopy` is extremely dangerous, and I want to save other users from a real footgun. If one naively calls `deepcopy(plan)` then it can lead to non-deterministic segfaults and extremely difficult to track Heisenbugs. Given that `deepcopy(plan)` is flawed, I think it should just error directly rather than return a memory-unsafe value.

**Comment by stevengj** [comment]
FFTW [plan execution is thread safe](https://www.fftw.org/fftw3_doc/Thread-safety.html), so it is perfectly fine to re-use the same plan across threads.
